### PR TITLE
[feature] atlantis integration

### LINKED ISSUE #278: atlantis support

### LINKED ISSUE #289: fogg generate aws/config for atlantis

### LINKED ISSUE #290: make targets/mode for Atlantis operations

### PR DESCRIPTION
**Pull request by ryanking**
This PR adds Atlantis (<runatlantis.io>) integration to fogg. The goal is to be able to use fogg to generate an atlantis.yaml configuration file and all related tools, scripts etc.

Happy to take feedback on both the approach and the implementation.

### Approach

The basic approach is, in no particular order–

* To the degree possible it would be great to be able to make changes to how Atlantis works without touching the Atlantis instances, so as much as possible is put into the per-repo config file.
* We allow configuring Atlantis on a per-component basis. I think this is important as I would expect teams to adopt it incrementally.
* We do a 2-phase plan - first . plan for each component (so that we can set a var in each) and then walk those plans and generate a plan for the atlantis.yaml and aws config
* We generate an aws config for atlantis. Because we rely on aws profiles for authentication, we need to find a way to get the Atlantis instance configured with all the right profiles. In the future we may want to think about doing this for other modes.
* If atlantis is enabled for a component, `make apply` will by default fail. You have to add `FORCE=1` to override. This should nudge people in the right direction while allowing a fall-back to running locally (important for adoption phase).
* we add a `make setup` that will install fogg and run `fogg setup` (you can run this from components, not just top-level)
* we no longer run `terraform get` since that is redundant with `terrraform init` and we **don't run -update=true anymore** which means that modules will not get updated. People should be using stable versions anyway and this makes it much faster
* tfenv is now installed inside the repo, rather than homedir. This makes it so that we can use tfenv in atlantis (rather than atlantis' own process for getting terraform versions)

--- 

Note there are also a handful of stylistic changes in this diff. Tried to keep those to a minimum, but also wanted to make some improvements, especially to our makefiles.

## Test Plan
* unit tests
* ran on shared-infra repo for 1 component

## References
* <https://www.runatlantis.io/docs/>
* Fixes #290 
* Fixes #289
* Fixes #278

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,10 +18,14 @@ lint: ## run the fast go linters
 	golangci-lint run
 .PHONY: lint
 
-packr: ## run the packr tool to generate our static files
+TEMPLATES := $(shell find templates -not -name "*.go")
+
+templates/a_templates-packr.go: $(TEMPLATES)
```
Comment (ryanking) [inline Makefile]: This *should* make it so that packr is only run if there are changed templates.

File: config/v2/config.go [hunk 2]
```diff
@@ -107,6 +114,7 @@ type SnowflakeProvider struct {
 }
 
 type Backend struct {
+	AccountID   *string `json:"account_id,omitempty"`
```
Comment (ryanking) [inline config/v2/config.go]: we need this to generate the aws config for atlantis

File: config/v2/validation.go [hunk 3]
```diff
@@ -173,6 +170,27 @@ func (c *Config) ValidateOktaProviders() error {
...
 }
 
+func (c *Config) ValidateAtlantis() error {
+	var errs *multierror.Error
+	c.WalkComponents(func(component string, comms ...Common) {
+		a := ResolveAtlantis(comms...)
+		// TODO refactor to a ResolveBackend method
+
+		// a can no longer be nil
+		if a.Enabled != nil && *a.Enabled {
```
Comment (edulop91) [inline config/v2/validation.go]: should we ditch enabled and instead go off the presence/absence or roleName?

File: config/v2/resolvers.go [hunk 4]
```diff
@@ -175,6 +176,24 @@ func ResolveTfLint(commons ...Common) v1.TfLint {
 	}
 }
 
+func ResolveAtlantis(commons ...Common) Atlantis {
```
Comment (edulop91) [inline config/v2/resolvers.go]: should it be `*Atlantis`? pattern matching against the other resolvers

File: config/v2/resolvers.go [hunk 5]
```diff
@@ -175,6 +176,24 @@ func ResolveTfLint(commons ...Common) v1.TfLint {
...
+		if c.Tools != nil && c.Tools.Atlantis != nil && c.Tools.Atlantis.Enabled != nil {
+			enabled = *c.Tools.Atlantis.Enabled
+		}
+	}
+
+	roleName := lastNonNil(AtlantisRoleNameGetter, commons...)
+	rolePath := lastNonNil(AtlantisRolePathGetter, commons...)
+
+	return Atlantis{
+		Enabled:  &enabled,
```
Comment (edulop91) [inline config/v2/resolvers.go]: should we ditch Enabled and instead go off the presence/absence of the atlantis block?

File: templates/common/fogg_header.tmpl [hunk 6]
```diff
@@ -0,0 +1,4 @@
+{{define "fogg_header"}}
```
Comment (ryanking) [inline templates/common/fogg_header.tmpl]: todo later, use this everywhere

File: plan/atlantis.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+	"fmt"
+	"sort"
+)
+
+type Atlantis struct {
+	Enabled  bool
+	Projects []AtlantisProject
+
+	// Profiles is a map of profile name -> role infro
+	Profiles map[string]AWSRole
```
Comment (edulop91) [inline plan/atlantis.go]: AWSProfiles maybe? I assume we might have to do similar things for other types of providers, should this be
```
Providers {
AWS: {}
}
```
to mirror how the v2 config specifies providers?

File: config/v2/resolvers.go (hunk 5 above)
Comment (edulop91) [reply config/v2/resolvers.go]: oh, I see. you need Enabled to properly resolve the inheritance chain

File: config/v2/resolvers.go (hunk 4 above)
Comment (ryanking) [reply config/v2/resolvers.go]: yeah, good catch

File: config/v2/resolvers.go (hunk 5 above)
Comment (ryanking) [reply config/v2/resolvers.go]: I don't think that would work in the case that you turn it on in defaults and want to turn it off somewhere else.

File: config/v2/validation.go [hunk 8]
```diff
@@ -64,10 +65,6 @@ func (c *Config) Validate() ([]string, error) {
 	return warnings, errs.ErrorOrNil()
 }
 
-func merge(warnings []string, err *multierror.Error, w []string, e error) ([]string, *multierror.Error) {
```
Comment (ryanking) [inline config/v2/validation.go]: was not in use

File: templates/repo/scripts/failed_output_only [hunk 9]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
```
Comment (ryanking) [inline templates/repo/scripts/failed_output_only]: I was trying to get this to work to suppress non-error output when running in atlantis. I didn't quite get it working, but going to leave it here to iterate on later.

File: templates/repo/scripts/component.mk [hunk 10]
```diff
@@ -8,41 +8,71 @@ include $(SELF_DIR)/common.mk
 all:
 .PHONY: all
 
+setup:
+	$(MAKE) -C $(REPO_ROOT) setup
```
Comment (edulop91) [inline templates/repo/scripts/component.mk]: .PHONY

File: templates/repo/scripts/component.mk [hunk 11]
```diff
@@ -8,41 +8,71 @@ include $(SELF_DIR)/common.mk
...
-.PHONY: get
-
-plan: terraform fmt get init
-	$(terraform_command) plan
+ifeq ($(MODE),local)
+plan: init fmt
+	@$(terraform_command) plan
+else ifeq ($(MODE),atlantis)
+plan: init lint
+	@$(terraform_command) plan -input=false -no-color -out $(PLANFILE) | scenery
```
Comment (edulop91) [inline templates/repo/scripts/component.mk]: what's `scenery`?

File: templates/repo/scripts/failed_output_only (hunk 9 above)
Comment (edulop91) [reply templates/repo/scripts/failed_output_only]: maybe leave a TODO on the file for future selves

File: testdata/bless_provider/scripts/component.mk [hunk 12]
```diff
@@ -8,41 +8,71 @@ include $(SELF_DIR)/common.mk
 all:
 .PHONY: all
 
+setup:
+	$(MAKE) -C $(REPO_ROOT) setup
```
Comment (edulop91) [inline testdata/bless_provider/scripts/component.mk]: does this need a .PHONY?

File: templates/repo/scripts/component.mk (hunk 11 above)
Comment (ryanking) [reply templates/repo/scripts/component.mk]: A terraform output prettifier written in go https://github.com/dmlittle/scenery

I have it baked into the atlantis image. Considering using fogg to install it for local mode too, but I want to see what happens with 0.12 output (it is supposedly better).

File: testdata/bless_provider/scripts/component.mk (hunk 12 above)
Comment (ryanking) [reply testdata/bless_provider/scripts/component.mk]: yeah, will get generated by fixiing above

**Pull request merged**